### PR TITLE
[FLINK-27269] Clean up the jar file after submitting the job

### PR DESCRIPTION
**Pull request by Aitozi**
This pr is meant to clean up the jar file after submitting the job to avoid the jar files exploding after serval upgrade

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java [hunk 1]
```diff
@@ -234,6 +238,22 @@ private JarUploadResponseBody jarUpload(FlinkSessionJob sessionJob, Configuratio
         }
     }
 
+    private void jarDelete(Configuration conf, String jarId) {
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:241]: nit: could we please change the name to `deleteJar`?

**Review by gyfora — APPROVED** [review]

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java (hunk 1 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java]: Fixed

**Comment by wangyang0918** [comment]
I am not sure whether deleting the submitted jars could actually work, which means freeing up the disk spaces. Because the classes are still in use in a dedicated user classloader. See https://github.com/apache/flink/blob/master/flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgram.java#L144.

I will have a manual verification soon.

**Comment by Aitozi** [comment]
@wangyang0918 Thanks for pointing this, I will also take a look then.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java [hunk 2]
```diff
@@ -234,6 +238,22 @@ private JarUploadResponseBody jarUpload(FlinkSessionJob sessionJob, Configuratio
...
+            JarDeleteHeaders headers = JarDeleteHeaders.getInstance();
+            JarDeleteMessageParameters parameters = headers.getUnresolvedMessageParameters();
+            parameters.jarIdPathParameter.resolve(jarId);
+            clusterClient
+                    .sendRequest(headers, parameters, EmptyRequestBody.getInstance())
+                    .get(
+                            operatorConfiguration.getFlinkClientTimeout().toSeconds(),
+                            TimeUnit.SECONDS);
+        } catch (Exception e) {
+            LOG.error("Failed to delete the jar.", e);
```
Comment (wangyang0918) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:253]: We need to print the `jarId` when failed.

**Review by wangyang0918 — APPROVED** [review]
Using `lsof | grep delete` in JobManager pod and do not find any leaked the jars after deletion. It seems that `userCodeClassLoader.close()` in `PackagedProgram#close()` has worked as expected.

+1 for merging.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java (hunk 2 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java]: Fixed

**Comment by Aitozi** [comment]
> Using `lsof | grep delete` in JobManager pod and do not find any leaked the jars after deletion. It seems that `userCodeClassLoader.close()` in `PackagedProgram#close()` has worked as expected.
> 
> +1 for merging.

👍🏻 Nice way to verify, thanks

**Pull request merged**